### PR TITLE
IGNITE-17125: Add dynamic completers

### PR DESCRIPTION
**Pull request by PakhomovAlexander**
`cluster/node config show/update` are dynamically suggested with configuration keys taken from REST.

https://issues.apache.org/jira/browse/IGNITE-17125

### CODE REVIEW

File: modules/cli/src/main/java/org/apache/ignite/cli/core/repl/completer/DynamicCompleter.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import java.util.List;
+
+/**
+ * Dynamic completer returns completions that can be fetched in real time
+ * from Ignite 3 node, file, or cached.
+ */
+public interface DynamicCompleter {
+
+    /**
+     * Given typed fords returns list of candidates that can be autocompleted.
```
Comment (valepakh) [inline modules/cli/src/main/java/org/apache/ignite/cli/core/repl/completer/DynamicCompleter.java:29]:
```suggestion
     * Given typed words returns list of candidates that can be autocompleted.
```

File: modules/cli/src/main/java/org/apache/ignite/cli/core/repl/completer/DynamicCompleterFilter.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+        return !(candidate.equals("--help") || candidate.equals("-h"));
+    }
+
+    private boolean filterClusterUrl(String[] words, String candidate) {
+        if (optionTyped(words)) {
+            return true;
+        }
+
+        return !session.isConnectedToNode() || (!candidate.equals("--cluster-url") && !candidate.equals("--node-url"));
```
Comment (valepakh) [inline modules/cli/src/main/java/org/apache/ignite/cli/core/repl/completer/DynamicCompleterFilter.java:62]: Todo - depending on what is merged first, this PR or the https://github.com/apache/ignite-3/pull/974, use `OptionsConstants.CLUSTER_URL_OPTION` here.

File: modules/cli/src/main/java/org/apache/ignite/cli/core/repl/completer/HoconDynamicCompleter.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    private final Config config;
+
+    /** Stores all keys that exist in the given config. Strings are stored in format: "root.subkey.subsubkey". */
+    private final Set<String> leafs;
+
+    /** List of all possible completions. */
+    private final List<String> completions;
+
+    /** Words, after those the completer should have been activated. */
+    private final Set<String> activationPostfixes;
```
Comment (valepakh) [inline modules/cli/src/main/java/org/apache/ignite/cli/core/repl/completer/HoconDynamicCompleter.java:42]: Should this be called activationPrefixes as in DynamicCompleterFactory?

File: modules/cli/src/main/java/org/apache/ignite/cli/core/repl/completer/HoconDynamicCompleter.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+                return words[i];
+            }
+        }
+        return "";
+    }
+
+    private void walkAndAdd(String keyPrefix, Set<String> keySet, List<String> result) {
+        keySet.forEach(key -> {
+            if (!leafs.contains(keyPrefix + key)) {
+                Set<String> nextKeySet = new HashSet<>(config.getConfig(keyPrefix + key).root().keySet());
```
Comment (valepakh) [inline modules/cli/src/main/java/org/apache/ignite/cli/core/repl/completer/HoconDynamicCompleter.java:100]: What's the point of wrapping the keySet in the new HashSet?

File: modules/cli/src/main/java/org/apache/ignite/cli/core/repl/completer/NodeUrlProvider.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        String urlInWords = findClusterUrlIn(words);
+        if (urlInWords != null) {
+            return urlInWords;
+        }
+
+        if (session.isConnectedToNode()) {
+            return session.nodeUrl();
+        }
+
+        return configManagerProvider.get().getCurrentProperty("ignite.cluster-url");
```
Comment (valepakh) [inline modules/cli/src/main/java/org/apache/ignite/cli/core/repl/completer/NodeUrlProvider.java:50]: We should use ConfigConstants.CLUSTER_URL here

File: modules/cli/src/main/java/org/apache/ignite/cli/core/repl/completer/NodeUrlProvider.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    private String findClusterUrlIn(String[] words) {
+        for (String word : words) {
+            if (word.startsWith("--cluster-url=")) {
+                return word.substring("--cluster-url=".length());
```
Comment (valepakh) [inline modules/cli/src/main/java/org/apache/ignite/cli/core/repl/completer/NodeUrlProvider.java:56]: And `OptionsConstants.CLUSTER_URL_OPTION` here

File: modules/cli/src/main/java/org/apache/ignite/cli/core/repl/executor/IgnitePicocliCommands.java [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+        CommandSpec spec = sub.getCommandSpec();
+        Help cmdhelp = new picocli.CommandLine.Help(spec);
+        List<AttributedString> main = new ArrayList<>();
+        Map<String, List<AttributedString>> options = new HashMap<>();
+        String synopsis = AttributedString.stripAnsi(spec.usageMessage().sectionMap().get("synopsis").render(cmdhelp));
+        main.add(HelpException.highlightSyntax(synopsis.trim(), HelpException.defaultStyle()));
+        // using JLine help highlight because the statement below does not work well...
+        //        main.add(new AttributedString(spec.usageMessage().sectionMap().get("synopsis").render(cmdhelp).toString()));
+        for (OptionSpec o : spec.options()) {
+            String key = Arrays.stream(o.names()).collect(Collectors.joining(" "));
```
Comment (valepakh) [inline modules/cli/src/main/java/org/apache/ignite/cli/core/repl/executor/IgnitePicocliCommands.java:132]: IDEA suggests replacing this with `String.join`

File: modules/cli/src/main/java/org/apache/ignite/cli/core/repl/executor/IgnitePicocliCommands.java [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+
+    /** {@inheritDoc} */
+    @Override
+    public List<String> commandInfo(String command) {
+        List<String> out = new ArrayList<>();
+        CommandSpec spec = cmd.getSubcommands().get(command).getCommandSpec();
+        Help cmdhelp = new picocli.CommandLine.Help(spec);
+        String description = AttributedString.stripAnsi(spec.usageMessage().sectionMap().get("description").render(cmdhelp));
+        out.addAll(Arrays.asList(description.split("\\r?\\n")));
+        return out;
```
Comment (valepakh) [inline modules/cli/src/main/java/org/apache/ignite/cli/core/repl/executor/IgnitePicocliCommands.java:153]:
```suggestion
        return Arrays.asList(description.split("\\r?\\n"));
```

Comment (valepakh) [reply modules/cli/src/main/java/org/apache/ignite/cli/core/repl/executor/IgnitePicocliCommands.java:153]: Or at least
```suggestion
        return new ArrayList<>(Arrays.asList(description.split("\\r?\\n")));
```

File: modules/cli/src/main/java/org/apache/ignite/cli/core/repl/executor/IgnitePicocliCommands.java [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+    }
+
+    @Override
+    public Object invoke(CommandRegistry.CommandSession session, String command, Object[] args) throws Exception {
+        List<String> arguments = new ArrayList<>();
+        arguments.add(command);
+        arguments.addAll(Arrays.stream(args).map(Object::toString).collect(Collectors.toList()));
```
Comment (valepakh) [inline modules/cli/src/main/java/org/apache/ignite/cli/core/repl/executor/IgnitePicocliCommands.java:160]:
```suggestion
        Arrays.stream(args).map(Object::toString).forEach(arguments::add);
```

File: modules/cli/src/main/java/org/apache/ignite/cli/core/repl/completer/HoconDynamicCompleter.java (hunk 3 above)
Comment (PakhomovAlexander) [reply modules/cli/src/main/java/org/apache/ignite/cli/core/repl/completer/HoconDynamicCompleter.java:42]: Well, maybe there is a better naming but they are different.

Let's look at the example
`cluster config show --cluster-urt=bleble --some-option --selector=rocksDb,`

`cluster config show` -- prefix

`--selector` --postfix

File: modules/cli/src/main/java/org/apache/ignite/cli/core/repl/completer/HoconDynamicCompleter.java (hunk 4 above)
Comment (PakhomovAlexander) [reply modules/cli/src/main/java/org/apache/ignite/cli/core/repl/completer/HoconDynamicCompleter.java:100]: You are right, there is no reason for that.

File: modules/cli/src/test/java/org/apache/ignite/cli/core/repl/completer/DynamicCompleterFilterTest.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+        // When
+        String[] filtered = new DynamicCompleterFilter(session).filter(words, candidates);
+
+        // Then cluster-url and help are filtered out
+        assertThat(asList(filtered), containsInAnyOrder("--selector"));
+    }
+
+    @Test
+    void doesNotFiltersHelpIfOptionIsTypedAndConnected() {
```
Comment (SammyVimes) [inline modules/cli/src/test/java/org/apache/ignite/cli/core/repl/completer/DynamicCompleterFilterTest.java:90]:
```suggestion
    void doesNotFilterHelpIfOptionIsTypedAndConnected() {
```

**Review by SammyVimes — CHANGES_REQUESTED** [review]
Good patch, just a few suggestions

**Review by SammyVimes — APPROVED** [review]
LGTM

**Pull request merged**